Lib/subprocess.py

```diff
@@ -1879,3 +1879,24 @@ def kill(self):
             """Kill the process with SIGKILL
             """
             self.send_signal(signal.SIGKILL)
+
+        @contextlib.contextmanager
+        def kill_on_error(self):
+            """Using context manager to kill subprocess if
+            a python exception is raised."""
+            try:
+                yield self
+            except:
+                if self.stdin:
+                    self.stdin.close()
+                if self.stdout:
+                    self.stdout.close()
+                if self.stderr:
+                    self.stderr.close()
+                try:
+                    self.kill()
+                except OSError:
+                    # process already terminated
```

Comment:
FYI - There is another reason why self.kill() can fail: PermissionError.  When a setuid process is launched (or potentially other system configurations enforcing process policies) you may get a PermissionError trying to signal that process.

When a PermissionError happens, the wait() call will block until the process dies on its own, if ever.  https://bugs.python.org/issue37091 covers this one.  It isn't clear if we should do anything for this situation.

Reply:
Looks try all possible exceptions or remove all possible exceptions is fine.